File: python/snake_game/ai/pathfinding.py
"""generic grid search (col, row) cells; blocked is a set. no snake/food knowledge."""
from collections import deque
import heapq


def in_bounds(cell, cols, rows):
    x, y = cell
    return 0 <= x < cols and 0 <= y < rows


def neighbors(cell, cols, rows):
    x, y = cell
    for dx, dy in ((1, 0), (-1, 0), (0, 1), (0, -1)):
        nxt = (x + dx, y + dy)
        if in_bounds(nxt, cols, rows):
            yield nxt


def _reconstruct(came_from, start, goal):
    path = [goal]
    while path[-1] != start:
        path.append(came_from[path[-1]])
    path.reverse()
    return path


def bfs_path(start, goal, blocked, cols, rows):
    """shortest path, or None if unreachable"""
    if start == goal:
        return [start]
    visited = {start}
    queue = deque([start])
    came_from = {}
    while queue:
        current = queue.popleft()
        for nxt in neighbors(current, cols, rows):
            if nxt in visited or nxt in blocked:
                continue
            visited.add(nxt)
            came_from[nxt] = current
            if nxt == goal:
                return _reconstruct(came_from, start, goal)
            queue.append(nxt)
    return None


def astar_path(start, goal, blocked, cols, rows):
    """manhattan-distance A*, or None if unreachable. same path length as bfs, fewer expansions."""
    def heuristic(a, b):
        return abs(a[0] - b[0]) + abs(a[1] - b[1])

    counter = 0  # tie-breaker so heap never compares cell tuples
    open_heap = [(heuristic(start, goal), 0, counter, start)]
    came_from = {}
    best_g = {start: 0}
    closed = set()

    while open_heap:
        _, g, _, current = heapq.heappop(open_heap)
        if current in closed:
            continue
        if current == goal:
            return _reconstruct(came_from, start, goal)
        closed.add(current)

        for nxt in neighbors(current, cols, rows):
            if nxt in blocked or nxt in closed:
                continue
            tentative_g = g + 1
            if tentative_g < best_g.get(nxt, float('inf')):
                best_g[nxt] = tentative_g
                came_from[nxt] = current
                counter += 1
                heapq.heappush(open_heap, (tentative_g + heuristic(nxt, goal), tentative_g, counter, nxt))
    return None


def flood_fill_count(start, blocked, cols, rows):
    """reachable cell count from start, ignoring blocked. proxy for available space."""
    if start in blocked:
        return 0
    visited = {start}
    queue = deque([start])
    count = 0
    while queue:
        current = queue.popleft()
        count += 1
        for nxt in neighbors(current, cols, rows):
            if nxt in visited or nxt in blocked:
                continue
            visited.add(nxt)
            queue.append(nxt)
    return count


def direction_between(a, b):
    return (b[0] - a[0], b[1] - a[1])

File: cpp/src/ai/pathfinding.cpp
#include "pathfinding.hpp"

#include <algorithm>
#include <array>
#include <cstdlib>
#include <queue>
#include <unordered_map>

namespace pathfinding {

bool InBounds(const Cell& c, int cols, int rows) {
    return c.col >= 0 && c.col < cols && c.row >= 0 && c.row < rows;
}

std::vector<Cell> Neighbors(const Cell& c, int cols, int rows) {
    static constexpr std::array<Cell, 4> kDeltas{Cell{1, 0}, Cell{-1, 0}, Cell{0, 1}, Cell{0, -1}};
    std::vector<Cell> result;
    result.reserve(4);
    for (const Cell& d : kDeltas) {
        Cell next{c.col + d.col, c.row + d.row};
        if (InBounds(next, cols, rows)) result.push_back(next);
    }
    return result;
}

namespace {

std::vector<Cell> Reconstruct(const std::unordered_map<Cell, Cell, CellHash>& cameFrom, const Cell& start,
                               const Cell& goal) {
    std::vector<Cell> path{goal};
    while (path.back() != start) {
        path.push_back(cameFrom.at(path.back()));
    }
    std::reverse(path.begin(), path.end());
    return path;
}

}  // namespace

std::optional<std::vector<Cell>> BfsPath(const Cell& start, const Cell& goal,
                                          const std::unordered_set<Cell, CellHash>& blocked, int cols,
                                          int rows) {
    if (start == goal) return std::vector<Cell>{start};

    std::unordered_set<Cell, CellHash> visited{start};
    std::queue<Cell> frontier;
    frontier.push(start);
    std::unordered_map<Cell, Cell, CellHash> cameFrom;

    while (!frontier.empty()) {
        Cell current = frontier.front();
        frontier.pop();
        for (const Cell& next : Neighbors(current, cols, rows)) {
            if (visited.count(next) || blocked.count(next)) continue;
            visited.insert(next);
            cameFrom[next] = current;
            if (next == goal) return Reconstruct(cameFrom, start, goal);
            frontier.push(next);
        }
    }
    return std::nullopt;
}

std::optional<std::vector<Cell>> AStarPath(const Cell& start, const Cell& goal,
                                            const std::unordered_set<Cell, CellHash>& blocked, int cols,
                                            int rows) {
    auto heuristic = [](const Cell& a, const Cell& b) {
        return std::abs(a.col - b.col) + std::abs(a.row - b.row);
    };

    struct Node {
        int f;
        int g;
        long counter;  // tie-breaker, avoids comparing Cells
        Cell cell;
    };
    struct Compare {
        bool operator()(const Node& a, const Node& b) const {
            if (a.f != b.f) return a.f > b.f;
            return a.counter > b.counter;
        }
    };

    std::priority_queue<Node, std::vector<Node>, Compare> open;
    std::unordered_map<Cell, Cell, CellHash> cameFrom;
    std::unordered_map<Cell, int, CellHash> bestG{{start, 0}};
    std::unordered_set<Cell, CellHash> closed;

    long counter = 0;
    open.push(Node{heuristic(start, goal), 0, counter++, start});

    while (!open.empty()) {
        Node current = open.top();
        open.pop();
        if (closed.count(current.cell)) continue;
        if (current.cell == goal) return Reconstruct(cameFrom, start, goal);
        closed.insert(current.cell);

        for (const Cell& next : Neighbors(current.cell, cols, rows)) {
            if (blocked.count(next) || closed.count(next)) continue;
            int tentativeG = current.g + 1;
            auto it = bestG.find(next);
            if (it == bestG.end() || tentativeG < it->second) {
                bestG[next] = tentativeG;
                cameFrom[next] = current.cell;
                open.push(Node{tentativeG + heuristic(next, goal), tentativeG, counter++, next});
            }
        }
    }
    return std::nullopt;
}

std::vector<Cell> FloodFillRegion(const Cell& start, const std::unordered_set<Cell, CellHash>& blocked,
                                   int cols, int rows) {
    if (blocked.count(start)) return {};

    std::unordered_set<Cell, CellHash> visited{start};
    std::queue<Cell> frontier;
    frontier.push(start);
    std::vector<Cell> region;

    while (!frontier.empty()) {
        Cell current = frontier.front();
        frontier.pop();
        region.push_back(current);
        for (const Cell& next : Neighbors(current, cols, rows)) {
            if (visited.count(next) || blocked.count(next)) continue;
            visited.insert(next);
            frontier.push(next);
        }
    }
    return region;
}

int FloodFillCount(const Cell& start, const std::unordered_set<Cell, CellHash>& blocked, int cols,
                    int rows) {
    return static_cast<int>(FloodFillRegion(start, blocked, cols, rows).size());
}

Direction DirectionBetween(const Cell& a, const Cell& b) {
    Cell delta{b.col - a.col, b.row - a.row};
    if (delta.col == 1) return Direction::Right;
    if (delta.col == -1) return Direction::Left;
    if (delta.row == 1) return Direction::Down;
    return Direction::Up;
}

}  // namespace pathfinding
